### ISSUE TITLE
Scatter_reduce onnx export error

### ISSUE THREAD

**Issue by sevenineone** [pyg-team/pytorch_geometric#8415]
I'm trying to export gnn model with onnx and get following error:

```bash
OnnxExporterError: ONNX does not support include_self=False for scatter_reduce
```

Here is example to reproduce this error:

```python
import torch
from torch import nn
from torch_geometric.nn import aggr, TransformerConv
from torch_geometric.nn.norm import GraphNorm


class GNN(nn.Module):
    def __init__(self):
        super(GNN, self).__init__()

        self.gnn = TransformerConv(10, 10, 1, dropout=0.1).jittable()         
        self.normalization = GraphNorm(in_channels=10)
        self.act = nn.GELU()
        self.aggr = aggr.AttentionalAggregation(gate_nn=nn.Linear(10, 1), nn=nn.Linear(10, 10)) #doesn't compile

    def forward(self, x, edge_index, batch_ptr):
        hidden = x
        hidden = self.gnn(hidden, edge_index)
        hidden = self.normalization(hidden)
        hidden = self.act(hidden)

        graph_rep = self.aggr(hidden, ptr=batch_ptr)

        return graph_rep

    
device = torch.device("cuda") if torch.cuda.is_available() else torch.device("cpu")
model = GNN().to(device)

x = torch.randn(100, 10).to(device)
edge_index = torch.randint(100, size=(2, 20)).to(device)
batch_ptr = torch.tensor([0, 100]).to(device)
res = model(x, edge_index, batch_ptr)
print(res)

torch.onnx.export(model, (x, edge_index, batch_ptr), 'model.onnx',
                  input_names=('x', 'edge_index', 'batch_ptr'), opset_version=18)
```

### Environment

* PyG version: 2.4.0 (master)
* PyTorch version: 2.1.0
* OS: ubuntu 22
* Python version: 3.10
* CUDA/cuDNN version: 12.3
* How you installed PyTorch and PyG (`conda`, `pip`, source): source

**Comment by rusty1s** [comment]
There is sadly not much we can do here besides waiting that `torch.scatter_reduce` is supported by ONNX :(

**Comment by sevenineone** [comment]
`torch.scatter_reduce` is supported by ONNX. ` include_self=False ` parameter is unsuported. Changing this to `True` will have almost no effect on the logic, but will make conversion to onnx possible

https://github.com/pyg-team/pytorch_geometric/blob/8833d0d473226f67c6880dc8c48caae8d0e2638d/torch_geometric/utils/scatter.py#L94-L95

**Comment by sevenineone** [comment]
Also `batch_ptr` is completely unsuported as it uses `segment_csr` from `torch_scatter`.
https://github.com/pyg-team/pytorch_geometric/blob/8833d0d473226f67c6880dc8c48caae8d0e2638d/torch_geometric/utils/segment.py#L23-L25
But using `index` makes conversion possible, as it uses `scatter`. (just a hint)
https://github.com/pyg-team/pytorch_geometric/blob/8833d0d473226f67c6880dc8c48caae8d0e2638d/torch_geometric/utils/softmax.py#L54-L77

**Comment by lee-clement-oxb** [comment]
I took a stab at addressing this in [#9587](https://github.com/pyg-team/pytorch_geometric/pull/9587). It's a little more complicated than simply switching the offending `scatter_reduce_` calls to `include_self=True` because they're applying min/max reductions and doing so would change the behaviour. I was able to implement a behaviour-preserving workaround that ONNX can support, but unfortunately it fails 4 tests related to `torch.dynamo` as it seems to introduce some graph breaks. I couldn't find a way through those issues, but maybe someone else can. In the meantime I'll use [#9587](https://github.com/pyg-team/pytorch_geometric/pull/9587) as a local patch since I'm not currently using `torch.dynamo`.
